blog/_src/posts/2019-05-14-scoping-in-r.md

```diff
@@ -0,0 +1,329 @@
+then evaluate it. This assigns `0` to `x`, which is no longer a free variable
+when it is used on line 4.
+
+#### Simulating dynamic scope
+
+```r
+x <- 1
+f <- function(a) {
+  get("x", envir = parent.frame()) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 2
+```
+
+On line 3, we perform an explicit variable lookup for `x`, but we do so in the
+environment `parent.frame()`, which refers to the calling function's
+environment.
+
+#### Constructing an arbitrary environment
+
+```r
+x <- 1
+f <- function(a) {
+  e <- new.env()
+  e$x <- 3
+  get("x", envir = e) + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # 3
+```
+
+On line 3, we construct an empty environment, `e`. Next, on line 4, we add a
+new binding to that environment, where `x` has the value `4`. Finally, on line
+5, we perform a variable lookup in environment `e`.
+
+#### Deleting bindings
+
+```r
+x <- 1
+f <- function(a) {
+  rm("x", envir = parent.env(environment()))
+  x + a
+}
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # Error in f(0) : object 'x' not found
+```
+
+On line 3, we delete the binding `x` from the given environment,
+`parent.env(environment())`. This refers to the parent (or enclosing)
+environment of the current environment of `f`, i.e. the environment where `f`
+was defined, which is the top-level environment.
+
+
+### What does "scoping in R is dynamic" mean?
+
+In _Optimizing R Language Execution via Aggressive Speculation_ (DLS 2016)
+([doi][swhj2016-doi]), Stadler, Welc, Humer, and Jordan write:
+
+> [...] variable scoping in R is dynamic and can be modified at the language
+> level [...]
+
+[swhj2016-doi]: https://doi.org/10.1145/2989225.2989236
+
+Our previous examples should demonstrate that scoping in R "can be modified at
+the language level."
+
+It is also true that "variable scoping in R is dynamic," in the sense that
+scoping is determined at run time. However, this is more general than the
+definition of _dynamic scoping_ we used in this blog post.
+
+Similarly, I've seen the term _lexical scoping_ used with a less specific but
+stronger definition---that all variable bindings can be determined at compile
+time. Under these more general definitions, R would be dynamically scoped
+rather than lexically scoped.
+
+I prefer the more specific definitions of lexical and dynamic scoping; I think
+they're more useful. Either way, it's important to be clear and make sure
+your audience knows what definitions you're using!
+
+---
+
+## Extra: another taxonomy for scoping rules
+
+In _[Lexical Scope and Statistical Computation][gi2000]_ (2000)
+([doi][gi2000-doi]), Gentleman and Ihaka, the creators of R, define four kinds
+of scoping rules for the purpose of their discussion:
+
+[gi2000]: https://www.stat.auckland.ac.nz/~ihaka/downloads/lexical.pdf
+[gi2000-doi]: https://doi.org/10.1080/10618600.2000.10474895
+
+  - *trivial*: no free variables allowed
+  - *dynamic*: as we defined it; free variables are evaluated in their dynamic
+     context
```

Comment:
So, "context" (in my opinion) is yet another synonym in the chain: scope — environment — context. I think having three words for essentially one technical concept makes the life of the reader way harder.

Reply:
This is an excellent point and I agree completely. I also like your suggestion above of explicitly defining environment.